src/irmin/tree.ml

```diff
@@ -1257,116 +1280,70 @@ module Make (P : Private.S) = struct
-      let* () =
-        match n.Node.v with
-        | Value (_, _, Some _) ->
-            let+ v = Node.to_value n >|= get_ok "export" in
-            n.v <- Value (repo, v, None)
-        | _ -> Lwt.return_unit
-      in
-      (* 2. push the current node job on the stack. *)
-      let () =
-        match (n.v, Node.cached_value n) with
-        | _, Some v -> Stack.push (add_node n v) todo
-        | Map x, None -> Stack.push (add_node_map n x) todo
-        | _ -> assert false
-      in
-      let contents = ref [] in
-      let nodes = ref [] in
-      List.iter
-        (function
-          | `Contents c -> contents := c :: !contents
-          | `Node n -> nodes := n :: !nodes)
-        l;
-
-      (* 2. push the contents job on the stack. *)
-      List.iter
-        (fun (c, _) ->
-          let h = Contents.hash c in
-          if Hashes.mem seen h then ()
-          else (
-            Hashes.add seen h ();
-            match c.Contents.v with
-            | Contents.Hash _ -> ()
-            | Contents.Value x -> Stack.push (add_contents c x) todo))
-        !contents;
-
-      (* 3. push the children jobs on the stack. *)
-      List.iter
-        (fun n ->
-          Stack.push (fun () -> (add_to_todo [@tailcall]) n Lwt.return) todo)
-        !nodes;
-      k ()
-    in
-
-    let rec loop () =
-      let task = try Some (Stack.pop todo) with Stack.Empty -> None in
-      match task with None -> Lwt.return_unit | Some t -> t () >>= loop
+    let rec on_node (`Node n) k =
+      match n.Node.v with
+      | Node.Hash (_, h) ->
+          (* Not modified *)
+          Node.export ?clear repo n h;
+          k ()
+      | Node.Value (_, v, None) ->
+          (* Not modified *)
+          let h = P.Node.Key.hash v in
+          Node.export ?clear repo n h;
+          k ()
+      | _ ->
+          let* children_seq =
+            let+ seq = Node.seq_unordered n in
+            Seq.map (fun (_, x) -> x) (get_ok "export" seq)
+          in
+          on_node_seq children_seq @@ fun () ->
+          let* v = Node.to_value n in
+          let v = get_ok "export" v in
+          let key = Node.hash n in
+          cnt.node_mem <- cnt.node_mem + 1;
+          let* mem = P.Node.mem node_t key in
+          let* () =
+            if mem then Lwt.return_unit
+            else (
+              cnt.node_add <- cnt.node_add + 1;
+              let+ key' = P.Node.add node_t v in
+              assert (equal_hash key key'))
+          in
+          Node.export ?clear repo n key;
+          k ()
+    and on_contents (`Contents (c, _)) k =
+      match c.Contents.v with
+      | Contents.Hash (_, key) ->
+          Contents.export ?clear repo c key;
+          k ()
+      | Contents.Value _ ->
+          let* v = Contents.to_value c in
+          let v = get_ok "export" v in
+          let key = Contents.hash c in
+          cnt.contents_mem <- cnt.contents_mem + 1;
+          let* mem = P.Contents.mem contents_t key in
```

Comment:
I've added this check, it doesn't exist on master. 

To do so, I had to change the signature of a polyvar in `Tree.export` and `Store.save_tree`

Reply:
not sure it's super useful as you `P.Contents.add` will do this anyway. As there is no recursion to stop here, I don't think that call wins us anything.